src/TraceEvent/TraceLog.cs

```diff
@@ -861,47 +861,68 @@ private unsafe void DispatchClonedEvent(TraceEvent toSend)
         /// </summary>
         internal void FlushRealTimeEvents(int minimumAgeMs = 0)
         {
-            lock (realTimeQueue)
+            // we need to guard our data structures from concurrent access.  TraceLog data
+            // is modified by this code as well as code in FlushRealTimeEvents.
+            if (!isFlushingRealTimeEvents)
             {
-                var nowTicks = Environment.TickCount;
-                // TODO review.
-                while (realTimeQueue.Count > 0)
+                lock (realTimeQueue)
                 {
-                    QueueEntry entry = realTimeQueue.Peek();
-                    // If it has been in the queue less than 1 second, we we wait until next time) & 3FFFFFF does wrap around subtraction.
-                    if (minimumAgeMs > 0 && ((nowTicks - entry.enqueueTick) & 0x3FFFFFFF) < minimumAgeMs)
+                    if (!isFlushingRealTimeEvents)
```

Comment:
If I'm reading this correctly, if you get here, then by definition `isFlushingRealTimeEvents` is false because the lock won't allow any callers into this section of the code when `isFlushingRealTimeEvents` is true.  I think you can get rid of the check inside the `lock` block.

Reply:
I have reread the relevant code. It seems that the purpose of this code is to ensure that any event is not dispatched until after minimumAgeMs. Assuming minimumAgeMs is 1000, an event enqueue at 900ms, and FlushRealTimeEvents is triggered at 1000ms. This code can guarantee that the event is postponed until the next invocation of FlushRealTimeEvents.

Reply:
Just to clarify, I'm referring to removing line 870 in the new diff.  If you get into the lock `isFlushingRealTimeEvents` is guaranteed to be false, and so you don't need to check it again, you can just set it to true and call `FlushRealtimeEventsNoLock`.

Reply:
The github unified diff view caused my misunderstanding 😂. Fixed.